File: javascript/lightsout/maven-selenium/lightsout/src/test/java/jp/co/teratech/intern/lightsout/test/util/SimultaneousEquation.java
package jp.co.teratech.intern.lightsout.test.util;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SimultaneousEquation {

	public static void main(String args[]) {
		// TODO: 動確
		long start = System.currentTimeMillis();

		Boolean[] b = {
			false,  true, false,  true, false,
			false, false, false, false, false,
			false, false, false, false, false,
			false, false, false, false, false,
			false, false, false, false, false,
		};

//		Boolean[] b = {
//			 true, false, false, false, false,
//			false, false, false, false, false,
//			false, false, false, false, false,
//			false, false, false, false, false,
//			false, false, false, false,  true,
//		};

		List<Boolean> board = Arrays.asList(b);
		SimultaneousEquation sim = new SimultaneousEquation();
		sim.init();                                     // 連立方程式を作成する
		List<Boolean> answer = sim.calculate(board);
		Util.dumpStatOfBoard(answer);
		System.out.println("Time: " + (System.currentTimeMillis() - start) + "ms");
	}

	/** 連立方程式 */
	Expressions simultaneous;

	/**
	 * 連立方程式を構築する
	 */
	public void init() {
		simultaneous = new Expressions();

		// 各ボタンに影響するボタンの式一覧を作成する
		for(int i = 0; i < Util.SIZE_OF_BOARD; ++i) {
			simultaneous.add(new Expression(i));
		}
		simultaneous.debugExpressions();
		simultaneous.formatExpressions();
	}

	/**
	 * 連立方程式から解を求める。解がない場合はnull を返す
	 * @param board 盤の状態
	 * @return 解。どのライトを押せばよいかマーキングされたBoolean リスト
	 */
	public List<Boolean> calculate(List<Boolean> board) {

		// 解があるかどうかを確認する
		for(int index: simultaneous.validationPoints) {
			Expression expression = simultaneous.expressions.get(index);
			boolean notExistence = false;

			for(int i = 0; i < expression.rightMember.size(); ++i) {
				if(expression.rightMember.get(i)) {
					notExistence = (notExistence != board.get(i));
				}
			}

			// ゼロな状態(false) であるかをチェックする。非zeroな状態(true) である場合は解無し
			if(notExistence) {
				return null;
			}
		}

		// 解ありと判定された場合は解を求める
		List<Boolean> result = getPlainBooleanList();
		for(int i : simultaneous.availablePoints) {
			if(board.get(i)) {  // TODO: 有効な式のみを格納したインデックスリストのようなものを作っておけば綺麗かも

				Expression expression = simultaneous.expressions.get(i);

				for(int j = 0; j < expression.rightMember.size(); ++j) {
					result.set(j, (result.get(j) != expression.rightMember.get(j)));
				}
			}
		}

		return result;
	}

	/**
	 * 初期化された盤(Boolean リスト) を取得する。
	 * @return 初期化された盤
	 */
	public static List<Boolean> getPlainBooleanList() {
		return SimultaneousEquation.getPlainBooleanList(false);
	}

	/**
	 * 引数で初期化された盤(Boolean リスト) を取得する。
	 * @param initializer 初期化値
	 * @return 初期化された盤
	 */
	public static List<Boolean> getPlainBooleanList(boolean initializer) {
		Boolean[] b = new Boolean[Util.SIZE_OF_BOARD];
		Arrays.fill(b, initializer);

		return Arrays.asList(b);
	}

	/**
	 * 連立方程式の式一覧
	 */
	static class Expressions {
		// TODO: 記述が面倒なのでprivate によるカプセル化は無しで...

		/** 連立方程式格納リスト */
		protected List<Expression> expressions = new ArrayList<Expression>();
		/** 無効な(有効とはいえない)式を持つマス目リスト。正解を求める事はできないが、解ありor無しを検証できるマス目 */
		protected List<Integer> validationPoints = new ArrayList<Integer>();
		/** 有効な式を持つマス目リスト */
		protected List<Integer> availablePoints  = new ArrayList<Integer>();

		/** 式を連立していく */
		public void add(Expression express) {
			expressions.add(express);
		}

		/**
		 * 各ボタン毎に正規化された連立方程式を作成する
		 */
		public void formatExpressions() {

			for(int i = 0; i < expressions.size(); ++i) {
				List<Expression> effected = getEffectedExpressionFromLeft(i);

				if(effected == null) {
					validationPoints.add(i);
					continue;
				};

				availablePoints.add(i);
				calcSystemOfEquations(effected, i);
			}

			// TODO: 完成した連立方程式をdebug 出力
			System.out.println("------------------------------------------------------");
			this.debugExpressions();
		}

		/**
		 * 連立方程式の解を解く
		 * @param expressions 連立方程式
		 * @return
		 */
		public void calcSystemOfEquations(List<Expression> effectedExpressions, int pivotIndex) {
			Expression pivot = expressions.get(pivotIndex);

			for(Expression effectedExpression : effectedExpressions) {
				// pivotIndex のExpression とeffectedExpression の排他論理和を求める。
				// 計算後の式は、effectedExpression 参照先に直接入れ込む
				for(int i = 0; i < effectedExpression.leftMember.size(); ++i) {
					effectedExpression.leftMember.set(i, effectedExpression.leftMember.get(i) != pivot.leftMember.get(i));
					effectedExpression.rightMember.set(i, effectedExpression.rightMember.get(i) != pivot.rightMember.get(i));
				}
			}
		}

		/**
		 * 必要に応じて式をswap する。
		 * 連立方程式のindex (引数)行目が、消去対象となる(index 番目の)ライトを未知数としていない場合
		 * 他の行にある消去対象となる(index 番目の)ライトを未知数としている式を
		 * index 行目の式と交換する。
		 * @param index 連立方程式の行数及び消去対象のライト
		 */
		public void swapExpressionsIfNessesally(int index) {
			if(expressions.get(index).leftMember.get(index) == false) {
				// 指定したインデックスと、ボタンのインデックスが異なる場合は、swap する。
				// 例) 2番目のExpression なのに、2番目のライトがtrue になっていない。
				//       -> 2番目のExpression と2番目以降のExpression で2番目のライトがtrue になっているExpression とswap する

				for(int i = index + 1; i < expressions.size(); ++i) {
					if(expressions.get(i).leftMember.get(index) == true) {
						Expression tmp = expressions.get(index);
						expressions.set(index, expressions.get(i));
						expressions.set(i, tmp);
						break;
					}
				}
			}
		}

		/**
		 * 連立方程式の解を求めるため、指定したライトの位置に影響のある式を返す。
		 * 影響のあり/なしは式の左辺から判断する。
		 * @param position ライトの位置
		 * @return 影響のある式のリスト
		 */
		public List<Expression> getEffectedExpressionFromLeft(int position) {

			// TODO: debug. swap 前に結果を出力してみる
			System.out.println("- dump the expression which might be swapped -");
			this.debugExpressions(position);

			swapExpressionsIfNessesally(position);

			// TODO: debug. 計算前に結果を出力してみる
			System.out.println("- dump the expression which has swapped -");
			this.debugExpressions(position);

			if(!expressions.get(position).leftMember.get(position)) {
				// 式がpivot として使えない(式の左辺に消去対象のライトが未知数になっていない)のであれば、
				// 連立方程式が成り立たない(連立方程式の未知数部分が無い==左辺が0)となるので
				// その式の計算は成り立たない。
				return null;
			}

			List<Expression> results = new ArrayList<Expression>();

			for(int i = 0; i < expressions.size(); ++i) {
				if(i == position) continue;

				if(expressions.get(i).leftMember.get(position)) {
					results.add(expressions.get(i));
				}
			}

			return (results.size() == 0 ? null: results);
		}

		/**
		 * 連立方程式の内容をデバッグ出力する
		 */
		public void debugExpressions() {
			debugExpressions(-1);
		}

		/**
		 * 連立方程式の内容をデバッグ出力する
		 * @param position マーカーを付けるポジション
		 */
		public void debugExpressions(int position) {
			Expression expression;
			for(int i = 0; i < expressions.size(); ++i) {

				System.out.print((i == position? "* ": "  "));
				System.out.printf("%04d:", i);

				expression = expressions.get(i);
				for(int j = 0; j < expression.leftMember.size(); ++j) {
					if(j % Util.WIDTH_OF_BOARD == 0) System.out.print(" ");
					System.out.print((expression.leftMember.get(j) ? "1": "0"));
				}
				System.out.print(" | ");
				for(int j = 0; j < expression.rightMember.size(); ++j) {
					if(j % Util.WIDTH_OF_BOARD == 0) System.out.print(" ");
					System.out.print((expression.rightMember.get(j) ? "1": "0"));
				}
				System.out.println();
			}
		}
	}

	/**
	 * 1 つの式
	 */
	static class Expression {
		// TODO: 記述が面倒なのでprivate によるカプセル化は無しで...

		/** 式のインデックス。何番目のボタンを解くために使う式か判別するためのインデックス */
		protected int index = -1;
		/** 連立方程式の左辺 */
		protected List<Boolean> leftMember = new ArrayList<Boolean>();
		/** 連立方程式の右辺 */
		protected List<Boolean> rightMember = new ArrayList<Boolean>();

		public Expression() {
			this(-1);
		}

		public Expression(int index) {
			this.index = index;
			for(int i = 0; i < Util.SIZE_OF_BOARD; ++i) {
				leftMember.add(false);
				rightMember.add(false);
			}

			if(index >= 0) {
				Util.getExpectedCheckStates(leftMember, index);
				rightMember.set(index, true);
			}
		}
	}

}

File: javascript/lightsout/development/js/simultaneous_equation.js
function SimultaneousEquation(sizeOfBoard) {
  var myself = this;
  /** 盤のライトの数 */
  const SIZE_OF_BOARD = sizeOfBoard || 25;
  /** 盤の縦横の数 */
  const WIDTH_OF_BOARD = Math.sqrt(SIZE_OF_BOARD);
  /** 連立方程式の左辺を取り出すインデックス値 */
  const I_EXPRESSION_LEFT = 0;
  /** 連立方程式の右辺を取り出すインデックス値 */
  const I_EXPRESSION_RIGHT = 1;
  /** 連立方程式 */
  var expressions;
  /** 解あり or 無しを判定するための方程式を取得するためのインデックス一覧 */
  var validatePoints = [];
  /** 解を求めるための方程式を取得するためのインデックス一覧 */
  var availablePoints = [];

  /** 連立方程式を作成して整理する */
  this.init = function() {
    expressions = new Array(SIZE_OF_BOARD);
    for(var i = 0; i < expressions.length; ++i) {
      var expressionLeft = new Array(SIZE_OF_BOARD);
      var expressionRight = new Array(SIZE_OF_BOARD);
      expressionLeft.fill(false);
      expressionRight.fill(false);

      myself.getExpectedStatus(expressionLeft, i);
      expressionRight[i] = true;
      expressions[i] = [expressionLeft, expressionRight];
    }

    // 連立方程式を各ボタンごとに整理された形式に変換する
    for(var i = 0; i < expressions.length; ++i) {
      var effected = myself._getEffectedExpressionFromLeft(i);

      if(effected === undefined) {
        validatePoints.push(i);
        continue;
      }

      availablePoints.push(i);
      myself._calcSystemOfEquations(effected, i);
    }

    // myself.dumpExpressions(expressions);  // TODO: debug: 完成した連立方程式を表示する
  }

  /**
   * 連立方程式を用いて解を求める
   * @param 盤の状態
   */
  this.calculate = function(board) {
    // 解があるかどうかを判定する。解が無い場合はreturn null
    var notExistence = false;

    for(var i = 0; i < validatePoints.length; ++i) {
      var right = expressions[validatePoints[i]][I_EXPRESSION_RIGHT]
      for(var j = 0; j < right.length; ++j) {
        if(right[j]) notExistence = (notExistence !== board[j]);
      }

      if(notExistence) return null;
    }

    // 解がある場合は解を求める
    var result = new Array(SIZE_OF_BOARD);
    result.fill(false);

    availablePoints.forEach((index) => {
      if(board[index]) {
        expression = expressions[index];
        for(var i = 0; i < expression[I_EXPRESSION_RIGHT].length; ++i) {
          result[i] = (result[i] !== expression[I_EXPRESSION_RIGHT][i]);
        }
      }
    });

    return result;
  }

  /**
   * 連立方程式を用いて、必ず解のある状態の配置でライトをシャッフルする
   */
  this.shuffle = function() {
    var result = new Array(SIZE_OF_BOARD);
    result.fill(false);
    var expression;

    for(var i = 0; i < availablePoints.length; ++i) {
      expression = expressions[availablePoints[i]][I_EXPRESSION_LEFT];

      if(Math.floor(Math.random() * 2) % 2 === 1) continue;
      for(var j = 0; j < SIZE_OF_BOARD; ++j) {
        if(expression[j]) {
          result[j] = !result[j];
        }
      }
    }

    return result;
  }

  /**
   * 連立方程式を計算する
   * @param effected 影響のある式一覧
   * @param pivotIndex 計算を行うときに元となる式
   */
  this._calcSystemOfEquations = function(effected, pivotIndex) {
    var pivot = expressions[pivotIndex];

    for(var i = 0; i < effected.length; ++i) {
      for(var j = 0; j < effected[i][I_EXPRESSION_LEFT].length; ++j) {
        effected[i][I_EXPRESSION_LEFT][j] = (effected[i][I_EXPRESSION_LEFT][j] != pivot[I_EXPRESSION_LEFT][j]);
        effected[i][I_EXPRESSION_RIGHT][j] = (effected[i][I_EXPRESSION_RIGHT][j] != pivot[I_EXPRESSION_RIGHT][j]);
      }
    }
  }

  /**
   * 連立方程式の中から、左辺の式から、引数に指定した位置のライトに影響のある式一覧を返す
   */
  this._getEffectedExpressionFromLeft = function(position) {
    myself._swapExpressionsIfNessesally(position);

    if(!expressions[position][I_EXPRESSION_LEFT][position]) {
      return undefined;
    }

    var results = [];

    for(var i = 0; i < expressions.length; ++i) {
      if(i === position) continue;

      if(expressions[i][I_EXPRESSION_LEFT][position]) {
        results.push(expressions[i]);
      }
    }

    return (results.length === 0? undefined: results);
  }

  /**
   * 連立方程式の式の順番を入れ替える
   * @param position 入れ替える式の場所
   */
  this._swapExpressionsIfNessesally = function(position) {
    if(!expressions[position][I_EXPRESSION_LEFT][position]) {
      for(var i = position + 1; i < expressions.length; ++i) {
        if(expressions[i][I_EXPRESSION_LEFT][position])  {
          var tmp = expressions[position];
          expressions[position] = expressions[i];
          expressions[i] = tmp;
          break;
        }
      }
    }
  }

  /**
   * あるボタンを押下したときの結果予想を取得する。
   * @param board 押下する盤の状態
   * @param position 押下位置
   */
  this.getExpectedStatus = function(board, position) {
    if(position > (WIDTH_OF_BOARD - 1)) {
      board[position - WIDTH_OF_BOARD] = !board[position - WIDTH_OF_BOARD];
    }
    if(position % WIDTH_OF_BOARD != 0) {
      board[position - 1] = !board[position - 1];
    }
    if((position % WIDTH_OF_BOARD) != (WIDTH_OF_BOARD - 1)) {
      board[position + 1] = !board[position + 1];
    }
    if(position < SIZE_OF_BOARD - WIDTH_OF_BOARD) {
      board[position + WIDTH_OF_BOARD] = !board[position + WIDTH_OF_BOARD];
    }

    board[position] = !board[position];
  }

  /**
   * 連立方程式をコンソール出力する
   * @param expressions 連立方程式の式一覧
   * @param position 出力時に行マークを付ける位置
   */
  this.dumpExpressions = function(expressions, position) {
    var left, right, line;
    var p = position || -1;

    for(var i = 0; i < expressions.length; ++i) {
      line = (p == i? "* ": "  ") + myself.getPadding4Digit(i) + i + ": ";

      left = expressions[i][I_EXPRESSION_LEFT];
      for(var j = 0; j < left.length; ++j) {
        line += (left[j] ? "1" : "0");
      }
      line += " | ";

      right = expressions[i][I_EXPRESSION_RIGHT];
      for(var j = 0; j < right.length; ++j) {
        line += (right[j] ? "1": "0");
      }
    }
  }

  /**
   * debug 出力のための数値パディング
   * @param num padding 対象の数字
   */
  this.getPadding4Digit = function(num) {
    if(num >= 1000)
      return "";
    if(num >= 100)
      return "0";
    if(num >= 10)
      return "00";

    return "000"
  }
}

